src/gccrs/args.rs

```diff
@@ -1,11 +1,85 @@
 //! This module interprets arguments given to `rustc` and transforms them into valid
 //! arguments for `gccrs`.
 
-use std::{env, path::PathBuf, process::Command};
+use std::convert::TryFrom;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+
+use getopts::Matches;
+
+use super::{EnvArgs, Error, Result, RustcOptions};
+
+/// A collection containing multiple instances of `Options`. This is necessary in order
+/// to circumvent the fact that `rustc` can currently generate multiple types of binaries
+/// with a single invokation.
+///
+/// For example,
+/// `rustc --crate-type=static --crate-type=dyn --crate-type=bin src/<file>.rs` will attempt
+/// to generate an executable, a dynamic library and a static one.
+///
+/// When using `gccrs` to compile things, we must also take into account what `gcc` can
+/// and cannot do. And `gcc` cannot currently generate multiple outputs in a single invokation.
+///
+/// Let's look at the previous example:
+///
+/// * To generate an executable from `src/<file>.rs`,
+/// we'd need to do the following: `gccrs src/<file>.rs` (with `-o <file>` if we don't
+/// want an executable named a.out, but that's not important).
+///
+/// * For a shared library, we need to add the `-shared` flag. On top of this, `rustc`
+/// generates libraries named `lib<name>.[so|a]` on Linux, while `gcc` will happily generate
+/// a shared library without any extension or prefix. This amounts to the following command:
+/// `gccrs -shared src/<file>.rs -o lib<file>.so`.
+///
+/// * Finally, `gcc` is not able to generate a static library at all. We *need* to use a
+/// different command, `ar`, in order to bundle up object files previously created by
+/// `gcc`. Therefore, we actually need *two* commands:
+/// `gccrs -c src/<file>.rs && ar csr src/<file>.o`
+///
+/// Multiple flags, such as `shared`, conflict with the generation of other binaries. On
+/// top of that, we cannot use the `-o` option precisely enough to control the output name
+/// based on the binary file produced.
+///
+/// Therefore, we need to wrap an unknown amount of sets of `gccrs` arguments in order to
+/// mimic a single `rustc` invokation. Later on, we need to iterate over those sets and
+/// spawn a new `gccrs` command for each of them.
+pub struct OptionsCollection {
+    args_set: Vec<Options>,
+}
```

Comment:
IIUC most of the `Options` are common to all `gccrs` invocations. So

```bash
rustc --crate-type=dyn --crate-type=bin src/<file>.rs --other-flag-1 --other-flag-2
```

becomes

```bash
gccrs src/<file>.rs --converted-other-flag-1 --converted-other-flag-2  # bin
gccrs -shared src/<file>.rs -o libfile.so --converted-other-flag-1 --converted-other-flag-2  # shared lib
```

where `--converted-other-flag-1 --converted-other-flag-2` are common to all invocations. So why not only store them once?

```rust
pub struct OptionsCollection {
    common_args: Vec<Arg>,
    crate_types: Vec<CrateType>,
    ... // Other fields that require multiple gccrs invocations, if any
}
```

I think that would make things way cleaner and would avoid saving the common options and input+output file name `N` times. 

Maybe even:

```rust
pub struct Args {
    input_files: Vec<String/PathBuf>,
    output_file: String/PathBuf,
    common_args: Vec<Arg>,
    crate_types: Vec<CrateType>,
    ... // Other fields that require multiple gccrs invocations, if any
}
```

Reply:
And only once you generate the `gccrs` cmmands, you will build the whole `Vec<Vec<String>>` argument matrix.

Reply:
As of right now, there aren't many shared options since the translator does not do much. I think your design makes a lot of sense for when we will handle all Rustc arguments (for example, after #46 is done). The only shared arguments for now are `-fPIE` and `-pie` on the `main` branch. I'm not against it but it's a bit hard to wrap my head around it with the current implementation. How do you feel about opening an issue and doing this in a later PR? If you think it's really important I'll try and do it now :)

Reply:
IMO simply the fact that you currently clone the input and output names N times would already warrant this design. And in the future there probably will be more shared arguments. So the current design, where you have to clone all of those arguments N times isn't scalable. 

Since this PR already improves the situation a bit, I'm good with merging this for now and to keep the technical debt.